### DISCUSSION TITLE
How best to support other developers

### DISCUSSION THREAD

**Discussion by andrew-sayers** [sleepdiary/conference-room#1] [Ideas]
Continuing a discussion from sleepdiary/dashboard#5 and sleepdiary/library#1, we should start thinking about how the project will be used by other real-world developers, and how to support them.

# Specific groups

Here are the groups I'm currently thinking about supporting.  It would be good to hear about any I've forgotten.

__People upgrading from Excel.__   A lot of people seem to like analysing their sleeping pattern in a spreadsheet.  Most of them will never want anything more than the dashboard's spreadsheet export option, but some of them might come here wanting to do some more detailed analysis.

__New developers.__  People generally discover their sleeping disorders at about the same time as they're thinking about careers, so it can be an obvious way to start exploring the possibility of a career in IT.

__Data analysts.__  People who analyse data for a living generally need to write a lot of code, but don't necessarily enjoy the act of programming.  They tend to gravitate towards Python, which supports that approach quite well.

__Node.js developers.__  A lot of our code is written in JavaScript, so it would be strange not to support people who already have experience in node

__Frontend developers.__  The primary interface to the project is on the web, so it seems reasonable to assume people with design ideas (who may or may not know much about coding) will want to give their two penneth.

# What should we do differently?

@quinn-dev has already suggested putting code into npm and providing specific instructions for node developers.  Those seem like good ideas, but we have to be careful not to let that outweigh other groups.

Data analysts in particular will probably be looking for a one-liner that lets them get data out of this program and into their own project.  @quinn-dev - would you be interested in writing a command-line interface to the library, so those people could do e.g. `sleepdiary my-sleep-as-android-diary.zip -to-standard > my-diary.json`?

The existing documentation more-or-less assumes the reader uses Linux and is familiar with containers.  New and frontend developers would probably be better served by a "getting started" document that walks them through the process of setting up a development environment, and maybe even choosing a text editor.  That's probably a job for me, as it's likely to involve bouncing around the codebase smoothing off sharp edges.

Finally, new developers will probably want inspiration for things to try out, and people upgrading from Excel will probably want how-to guides for specific things they want to achieve.  The "tutorials" section in [the library documentation](https://sleepdiary.github.io/library/doc/) tries to do that, but it's become quite well hidden and could do with updating.  If we put that sort of thing in a wiki, might people be willing to add things?

**Comment by quinn-dev** [comment]
First of all, writing a command-line interface to the library was exactly what I wanted to do as my first project using this library ^^
It arguably is a very obvious use case and makes the library useful for people who just want to do some data format conversion without fiddling with node.

But back to the main topic: I think that, in order to make the project more accessible to all of the above-mentioned groups, it would be a good idea to make it easier for people to find what they are looking for.
To that end, I propose having a reworked landing page, with an overview of the different parts of this project and prominent links to the "Getting started" section as well as the dashboard, as well as an improved documentation.
The current [library documentation](https://sleepdiary.github.io/library/doc/) is essentially a reference for the library's API in my opinion, and I'd treat it as such: A very technical document detailing the library, for tech-savvy people.
Everything else I'd put in a separate documentation, written in Markdown and deployed using something like [Docusaurus](https://docusaurus.io/) (which can also be used for the landing page).

My goal would be to have all the required information in there and just point people to the website.
What do you think about this?

**Reply by andrew-sayers** [reply]
I've made a note to rethink the documentation.  That'll be this weekend at the earliest, and I'll presumably have more to say after that.

Just to make sure we're talking about the same thing - the global landing page is https://sleepdiary.github.io/ - I assume you're referring to the  destination of the "source code" link?  That currently points to https://github.com/sleepdiary/, which is the best solution available right now, but yes we could do with something better.

An alternative destination for that link was https://sleepdiary.github.io/library/.  I don't think that's linked from anywhere right now, but it should probably be linked any new developer page.

https://sleepdiary.github.io/library/ is generated with GitHub Pages' internal markdown-based page builder.  I used the "midnight" theme because there seems to be an emerging standard that sleep apps use dark themes.  I'm not necessarily opposed to switching to e.g. Docusaurus, but it would need to be similarly themeable and have some other relevant benefit.

I think we're talking about putting new documentation in [the website repository](https://github.com/sleepdiary/sleepdiary.github.io) instead of e.g. a wiki or some new "documentation" repo?

**Reply by quinn-dev** [reply]
The main benefit of Docusaurus, in my opinion, would be drastic visual design improvements. It would be rather easy to create a nice-looking landing page, and I would argue that both the layout and the design of the actual documentation make it significantly more readable, at least when comparing the Docusaurus default theme with the "midnight" theme currently used by the documentation. As for theming and dark mode: A dark mode is already integrated, users can switch between a light and a dark theme. It's also possible to create a custom theme based on the default theme. You can take a look at the default theme [here](https://docusaurus.io/docs).

As for the destination of the docs: I think it would make most sense to have a landing page for the project at https://sleepdiary.github.io/, though this would mean we would have to place the web app elsewhere.

If I find the time by next week, I will work on a more concrete restructuring proposal to make my idea a bit more clear.

**Reply by andrew-sayers** [reply]
Looking back over this conversation, it seems like you're better-placed to lead the documentation work - writing a CLI will throw up all the right questions, and from the perspective of someone learning it rather than someone that's already internalised the library's assumptions.  Plus it sounds like you've got a strong concept for a layout.

On the other hand, I think I'm better-placed to redesign the development environment so people can get started more easily.  That will lead on naturally from the work I'm currently doing with the build process, so I'll add that to my todo list and let you know how I get on.

**Comment by andrew-sayers** [comment]
Here's a summary of developer-related improvements from this week:

I've added some [project-wide issue templates](https://github.com/sleepdiary/.github/tree/main/.github/ISSUE_TEMPLATE) and linked to them from most repos' README files (see e.g. [the dashboard's README](https://github.com/sleepdiary/dashboard).  I've also added a [support document](https://github.com/sleepdiary/.github/blob/main/SUPPORT.md) that should pop up the first time you create an issue.  Hopefully these will make things a bit less confusing for newcomers.

I've also created a new development server and a [getting started guide](https://github.com/sleepdiary/docs/blob/main/development/getting-started.md), which should reduce the number of hoops newbies have to jump through.  It should also improve performance for experienced developers - for example, the markdown renderer automatically refreshes the page when you save, which has been extremely helpful with all the documentation I've written today.

When you create a PR, a [pull request checklist](https://github.com/sleepdiary/.github/blob/main/pull_request_template.md) should automatically be pasted into the PR message.  This will hopefully give us a place to gather common problems and solutions.

I'm continuing to work on automatic tests that run on every PR.  The dev server includes a page that lets you run those tests privately, which will hopefully make it easier to avoid publicising code that fails tests.

I've created [a new "docs" repo](https://github.com/sleepdiary/docs) for documentation, so I can manage push access to documentation separately from code.  Speaking of which, @quinn-dev I've added you to the new "Documenters" team, which should let you merge PRs directly into [.github](https://github.com/sleepdiary/.github) and [docs](https://github.com/sleepdiary/docs).

I'll probably spend a lot of next week adding things to the "docs" repo, including moving [the info wiki](https://github.com/sleepdiary/info/wiki) in there.  GitHub wikis don't encourage participation strongly enough to make up for how ugly they are, and for example [MDN's block of helpful links](https://developer.mozilla.org/en-US/docs/Web#on-github) at the bottom of each page give you most of the benefits of a wiki with few of the costs.  I'd also like to try making all user pages dark by default and all developer pages light by default, so people immediately know when they've clicked on the wrong link.  I'll try doing all that with Docusaurus, but might switch if it doen't want to co-operate.

Reading back over the discussion of where to put documentation, I suspect we've been talking past each other.  I expect most people will come to the site wanting to use the dashboard (or other apps we might add), so it makes sense to use the same visual language on the landing page as in the apps.  But that page is currently implemented as [a separate Vue app](https://github.com/sleepdiary/sleepdiary.github.io/blob/main/index.html) that's just styled to look like the dashboard, so we can change it without any kind of structural upheaval.  I'm very open to a discussion of which links should be there and how prominent they should be, but I suspect I'll have stronger opinions when the docs repo is a bit more mature.  In the mean time, you might like to check out [vuetify's example pages](https://vuetifyjs.com/en/getting-started/wireframes/) for some ideas.

Despite taking all week, most of the above are just at the "minimal viable product" stage.  Hopefully they'll mature naturally with time, but any improvements you can think of would be very welcome :)

**Comment by andrew-sayers** [comment]
I guess you've lost interest in this project - that's fine, there've been plenty of projects I wanted to succeed but couldn't sustain motivation for.  But I'm about to do some housekeeping, so I wanted to make it clear why you're about to get a flurry of notifications...

[The new docs repo](https://sleepdiary.github.io/docs/) is now out - Docusaurus was definitely an improvement, but I went with [VuePress v2](https://v2.vuepress.vuejs.org/) in the end.  As the name suggests, it uses Vue - the same framework as the rest of the site.  That makes it easier to add interactive elements, which turns out to be worth the sometimes-clunky configuration.

I've made a new [conference-room repository](https://github.com/sleepdiary/conference-room) for issues that aren't associated with any particular repo.  I'll move our discussion threads over there, which you'll presumably be notified about somehow.

GitHub recommends creating "good first issue"s to help developers get started with a project, and some of the issues you've raised (e.g. moving away from CDNs) would be perfect for that.  I'll either move the existing issues to the conference room and rewrite them, or close them and create new ones - either way, more notifications.  You're welcome to pick the issues back up in future, otherwise they'll stay there waiting to help out the next new developer :)

And some of the things we talked about have been folded into other development work.  For example, ongoing process improvements should have made `yarn` work by now, and will eventually lead to publishing a library to npm.  I'll leave the relevant issues open as a reminder until the relevant work is all wrapped up.

Anyway, your contributions have set the project off in a new direction, so thank you and good luck in the future.
